### DISCUSSION TITLE
Model struggles with cumulative count data

### DISCUSSION THREAD

**Discussion by nsgoneape** [amazon-science/chronos-forecasting#158] [Q&A] [answered]
Hey Hello, I got the model running locally and it works very well with sample data. However when I pass in my application specific row count time series data, I get wonky unexpected results. I am trying to figure out why this is occuring.

**Comment by nsgoneape** [comment]
When I use this sample data C:\ONEAPE\iCEDQ\icedq.2\engines\forecast_engine\data\sorted_time_series_july_modified.csv
https://drive.google.com/file/d/1GDS3lZRl9khL8Crj8e7hfQNlZzvixDXw/view?usp=sharing
I get squirely and unexpected result.

![image](https://github.com/user-attachments/assets/a079bd04-c4fa-4cb5-a600-fca7173a1d69)

However, when I use data with oscillation, it seems to work a lot better.  https://drive.google.com/file/d/1lJzPCAWI03EX-12SxqgnBiOp3OOCi2n9/view?usp=sharing

![image](https://github.com/user-attachments/assets/1db9ded0-bb24-465d-9b44-f34698d08b9f)

Here is my code for context:
```python
import numpy as np
import pandas as pd
import torch
from chronos import ChronosPipeline
import plotly.graph_objects as go

# Load the ChronosPipeline
pipeline = ChronosPipeline.from_pretrained(
    "amazon/chronos-t5-large",
    device_map="cuda",
    torch_dtype=torch.bfloat16,
)

# Read the CSV file
csv_path3 = r"C:\ONEAPE\iCEDQ\icedq.2\engines\forecast_engine\data\Sales - Sales.csv"
df = pd.read_csv(csv_path3)

# Prepare the context and make predictions
context = torch.tensor(df["target"])
prediction_length = 50
forecast = pipeline.predict(context, prediction_length)

# Calculate forecast statistics
forecast_index = list(range(len(df), len(df) + prediction_length))  # Convert to list
low, median, high = np.quantile(forecast[0].numpy(), [0.1, 0.5, 0.9], axis=0)

# Create the Plotly figure
fig = go.Figure()

# Add historical data
fig.add_trace(go.Scatter(
    x=list(df.index),  # Convert to list
    y=df["target"],
    mode='lines',
    name='Historical Data',
    line=dict(color='royalblue')
))

# Add median forecast
fig.add_trace(go.Scatter(
    x=forecast_index,
    y=median,
    mode='lines',
    name='Median Forecast',
    line=dict(color='tomato')
))

# Add prediction interval
fig.add_trace(go.Scatter(
    x=forecast_index + forecast_index[::-1],
    y=high.tolist() + low.tolist()[::-1],
    fill='toself',
    fillcolor='rgba(255,99,71,0.3)',
    line=dict(color='rgba(255,99,71,0)'),
    name='80% Prediction Interval'
))

# Update layout
fig.update_layout(
    title='Time Series Forecast',
    xaxis_title='Time',
    yaxis_title='Value',
    legend_title='Legend',
    hovermode='x unified'
)

# Add gridlines
fig.update_xaxes(showgrid=True, gridwidth=1, gridcolor='lightgray')
fig.update_yaxes(showgrid=True, gridwidth=1, gridcolor='lightgray')

# Show the interactive plot
fig.show()
```

**Comment by lostella** [comment, marked as answer]
@nsgoneape the issue with your first series, is that its variance is rather small compared to its mean. This makes the model run into some loss of precision issues (see Sec. 5.7 in the [paper](https://arxiv.org/pdf/2403.07815), especially Fig. 16(b)).

Simply shifting the series down by it's minimum value, and applying the opposite shift to the resulting forecasts, recovers the intended behaviour in this case. See my updated snippet:

```
import numpy as np
import pandas as pd
import torch
from chronos import ChronosPipeline
import matplotlib.pyplot as plt

# Load the ChronosPipeline
pipeline = ChronosPipeline.from_pretrained(
    "amazon/chronos-t5-tiny",  # using tiny model for local testing
    # device_map="cuda",
    # torch_dtype=torch.bfloat16,
)

# Read the CSV file
csv_path3 = "/Users/stellalo/chronos-forecasting/scratchspace/sorted_time_series_july_modified.csv"
df = pd.read_csv(csv_path3)

# Prepare the context and make predictions
context = torch.tensor(df["target"])
prediction_length = 50

# NOTE: pre-processing and post-processing happen here
forecast = pipeline.predict(context - context.min(), prediction_length) + context.min()

# Calculate forecast statistics
forecast_index = list(range(len(df), len(df) + prediction_length))  # Convert to list
low, median, high = np.quantile(forecast[0].numpy(), [0.1, 0.5, 0.9], axis=0)

plt.plot(range(len(context)), context)
plt.plot(range(len(context), len(context) + prediction_length), median)
```

![image](https://github.com/user-attachments/assets/b017b129-996b-4e65-8597-5bd42955156c)

_Edit: expanding the answer_

One option to enforce monotonicity is to model the differenced series, using `.diff` for differentiating and `.cumsum` to integrate back the forecasted samples:

```
forecast = pipeline.predict(
    context.diff(dim=0),
    prediction_length
).cumsum(dim=-1) + context[-1]
```

This is what I get with `amazon/chronos-t5-large` in this case:

![image](https://github.com/user-attachments/assets/5c819df5-21a1-4c92-9964-53632b60b548)

**Comment by nsgoneape** [comment]
This great. The solution works with the tiny model. However, and this is interesting, it does not seems to perform as well with the large model.

Here are the two results, the only difference being the model choice.

This is with chronos-t5-large
![newplot](https://github.com/user-attachments/assets/9bbb94f6-b0b4-4477-982b-a10174376c7e)

This is with chronos-t5-small
![newplot (2)](https://github.com/user-attachments/assets/97c0785e-94ff-40fd-b642-222cdcdd7a1b)

As this is rowcount data, it will always increase linearly.

**Comment by lostella** [comment]
@nsgoneape one option to enforce monotonicity is to model the differenced series, using `.diff` for differentiating and `.cumsum` to integrate back the forecasted samples:

```
forecast = pipeline.predict(
    context.diff(dim=0),
    prediction_length
).cumsum(dim=-1) + context[-1]
```

This is what I get with `amazon/chronos-t5-large` in this case:

![image](https://github.com/user-attachments/assets/5c819df5-21a1-4c92-9964-53632b60b548)

**Reply by nsgoneape** [reply]
Wow, very nice, I will check this out when I get back to my computer.
